### PR TITLE
Advanced GPU Documentation

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Introduces a new advanced multi-gpu section, with more explanation and details. Cleanup of old APIs + addition of Fully Sharded and activation checkpointing.

A lot of the high level points may need actual data to back them up, but are collations from the DeepSpeed/FairScale team. It's more important right now to highlight the high level points, and then trickle down to data points via visualizations if possible.

If anyone has any suggestions on better naming than `Advanced GPU Optimized Training` Let me know! 

cc @ananthsub @shuyingsunshine21 @min-xu-ai

TODO:

- [x] Add DDP Communication Hooks section
- [x] Potentially come up with a better name than `Memory Optimized Multi-GPU Training` (how about `Advanced Multi-GPU Training`)?
- [x] Add FSDP requirement that you must use `self.trainer.model` when doing `configure_optimizers`

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+
+Some of these memory efficient plugins rely on offloading onto other forms of memory, such as CPU RAM or NVMe. This means you can even see memory benefits on a **single GPU**, using a plugin such as :ref:`deepspeed-zero-stage-3-offload`.
+
+Choosing an Optimized Multi-GPU Plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently all Memory Optimized Multi-GPU plugins shard the model states across your GPUs; just in different ways.
+
+This means as you scale up the number of GPUs, you can reach the number of model parameters you'd like to train.
+
```
Comment (ananthsub) [inline docs/source/advanced/optimized_multi_gpu.rst:19]: should we could also highlight DDP communication hooks?

Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst:19]: Yes!!! Great idea :)

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+.. code-block:: bash
+
+    PL_DEEPSPEED_CONFIG_PATH=/path/to/deepspeed_config.json python train.py --plugins deepspeed
+
+
+----------
+
+.. _sequential-parallelism:
+
+Sequential Model Parallelism with Checkpointing
```
Comment (min-xu-ai) [inline docs/source/advanced/optimized_multi_gpu.rst:595]:
```suggestion
Sequential Model Parallelism with Activation Checkpoint
```

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+PyTorch Lightning integration for Sequential Model Parallelism using `FairScale <https://github.com/facebookresearch/fairscale>`_.
+Sequential Model Parallelism splits a sequential module onto multiple GPUs, reducing peak GPU memory requirements substantially.
+We also provide auto-balancing techniques through FairScale, to find optimal balances for the model across GPUs.
+In addition, we use Gradient Checkpointing to reduce GPU memory requirements further, and micro-batches to minimizing device under-utilization automatically.
```
Comment (min-xu-ai) [inline docs/source/advanced/optimized_multi_gpu.rst:600]:
```suggestion
In addition, we use Activation Checkpoint (aka Gradient Checkpointing) to reduce GPU memory requirements further, and micro-batches to minimizing device under-utilization automatically.
```

File: docs/source/advanced/optimized_multi_gpu.rst (hunk 2 above)
Comment (min-xu-ai) [reply docs/source/advanced/optimized_multi_gpu.rst:595]: Is this also called pipelining?

**Review by min-xu-ai — APPROVED** [review]
fairscale part looks great to me. Thanks for adding this great doc!

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 4]
```diff
@@ -0,0 +1,661 @@
+Memory Optimized Multi-GPU Training
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst:1]: I've gone with `Advanced GPU Optimized Training` which might be nicer!

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+
+.. _sharded:
+
+Sharded Training
+^^^^^^^^^^^^^^^^
+Lightning integration of optimizer sharded training provided by `FairScale <https://github.com/facebookresearch/fairscale>`_.
+The technique can be found within `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
+`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_,
+however the implementation is built from the ground up to be pytorch compatible and standalone.
+Sharded Training allows you to maintain GPU scaling efficiency, whilst reducing memory overhead drastically. In short, expect normal linear scaling, and significantly reduced memory usage when training large models.
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply docs/source/advanced/optimized_multi_gpu.rst:51]: sorry, I meant "near-linear" above.

File: docs/source/advanced/optimized_multi_gpu.rst (hunk 2 above)
Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst:595]: This will be removed soon; I think we'll be deprecating this part of the docs and removing it + the implementation for now!

**Comment by SeanNaren** [comment]
Appreciate the comments @min-xu-ai will address the last comments ASAP!

File: docs/source/advanced/optimized_multi_gpu.rst [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+.. _manual-wrap:
+
+Manual Wrap
+"""""""""""
+
+To activate parameter sharding, you can also wrap layers using provided ``wrap`` or ``auto_wrap`` functions as described below.
+
+When not using Fully Sharded these wrap functions are a no-op. This means once the changes have been made, there is no need to remove the changes for other plugins.
+
+This is a requirement for really large models and also saves on instantiation time as modules are sharded instantly, rather than after the entire model is created in memory.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst:126]: Thanks @min-xu-ai, could you go into more details as to what is different here compared to DDP?

[the remark this answers is not in the document]
Comment (min-xu-ai) [reply docs/source/advanced/optimized_multi_gpu.rst:141]: Sure. With DDP, you can have this:
```
rank 0               rank 1
m = model()    m = model()  <------ two ranks may have different weights due to different random seeds
train(m)            train(m)  <---------------- weights are synced by ddp
```

With FSDP, since `m` is sharded, parts of the weights will be from rank 0 and parts of the weights will be from rank 1 when sharding happens. That can break the weight init assumptions, like zero mean and unit stddev etc.

Therefore, until FSDP can sync weights between ranks, weight init needs to be very careful with FSDP.

**Comment by williamFalcon** [comment]
maybe an advanced tutorials section?
and add this as the first bullet

**Comment by SeanNaren** [comment]
> maybe an advanced tutorials section?
> and add this as the first bullet

Agreed I actually plan on doing something similar in terms of layout to this which would be closer to an actual tutorial: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html

Like if I wanted to train a transformer model on my data, what are the steps I should take? What should the process look like?

**Review by ananthsub — COMMENTED** [review]
@SeanNaren another setting from DDP to enable memory savings is to set `gradient_as_bucket_view=True` : https://pytorch.org/docs/master/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel

This should save an extra ~10-15% of peak memory usage and can be an intermediary option for users who don't need the sharded/fully sharded/deepspeed enginees

cc @zhaojuanmao

File: docs/source/advanced/optimized_multi_gpu.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst:141]: What would be a potential solution here? use torch.distributed communications to sync global stats across all shards when initializing the model? It would be good to have a solution in place in the docs for users to have an example!

**Comment by SeanNaren** [comment]
I have added both the DDP Comm hooks (need to test it myself) + the `gradient_as_bucket_view` into the docs! Let me know what you think @ananthsub :)

Once FSDP is merged, we can merge this PR

**Comment by ananthsub** [comment]
@SeanNaren any harm in merging this now (minus FSDP) now and then adding FSDP once #6152 is merged?

**Comment by SeanNaren** [comment]
> @SeanNaren these are fantastic docs!!! these would be super useful to even merge now (minus FSDP) and then we can add back the FSDP section once #6152 is merged. What do you think?

Thanks @ananthsub let me remove the FSDP stuff and merge :) Was hoping we'll get the FSDP stuff merged, but will separate out to get this in ASAP

**Comment by SeanNaren** [comment]
I have as followups to this PR:

- Once FSDP merged, re-add FSDP docs
- Create advanced tutorial section and add this as a bullet point
- Once DeepSpeed infinity is merged, add this to the docs

File: docs/source/advanced/optimized_multi_gpu.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/advanced/optimized_multi_gpu.rst]: Just following up here because it might be a solution @min-xu-ai, but using `SummonFullParams` may be a way to init the model locally if the model would fit into memory, and then broadcast results. I'll add this into the FSDP docs in time when we merge the feature in!

**Comment by awaelchli** [comment]
Great docs @SeanNaren very high quality like everything you do.

Some sections have larger separation than others. Maybe it was intentional but I don't see the pattern. 

image

File: docs/source/advanced/advanced_gpu.rst [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+
+    from pytorch_lightning import Trainer
+    from fairscale.nn import checkpoint_wrapper
+
+
+    class MyModel(pl.LightningModule):
+        def __init__(self):
+            # Wrap layer using checkpoint_wrapper
+            linear_layer = checkpoint_wrapper(nn.Linear(32, 32))
+            self.block = nn.Sequential(linear_layer, nn.ReLU())
```
Comment (justusschock) [inline docs/source/advanced/advanced_gpu.rst:99]: would it make sense to either checkpoint the ReLU as well or checkpoint the whole block? The ReLU should be even cheaper to compute and free the same amount of memory.

**Review by justusschock — APPROVED** [review]
Great Work!

**Review by tchaton — APPROVED** [review]
Awesome work !

File: docs/source/benchmarking/performance.rst [hunk 8]
```diff
@@ -132,30 +132,11 @@ However, know that 16-bit and multi-processing (any DDP) can have issues. Here a
...
-
-----------
-
-Sequential Model Parallelism with Checkpointing
------------------------------------------------
-PyTorch Lightning integration for Sequential Model Parallelism using `FairScale <https://github.com/facebookresearch/fairscale>`_.
-Sequential Model Parallelism splits a sequential module onto multiple GPUs, reducing peak GPU memory requirements substantially.
-
-For more information, refer to :ref:`sequential-parallelism`.
+When training on single or multiple GPU machines, Lightning offers a host of advanced optimizations to improve throughput, memory effeciency and model scaling.
```
Comment (carmocca) [inline docs/source/benchmarking/performance.rst:138]:
```suggestion
When training on single or multiple GPU machines, Lightning offers a host of advanced optimizations to improve throughput, memory efficiency, and model scaling.
```

File: docs/source/advanced/training_tricks.rst [hunk 9]
```diff
@@ -149,9 +149,8 @@ The algorithm in short works by:
...
 
-Sequential Model Parallelism with Checkpointing
----------------------------------------------------------------------
-PyTorch Lightning integration for Sequential Model Parallelism using `FairScale <https://github.com/facebookresearch/fairscale>`_.
-Sequential Model Parallelism splits a sequential module onto multiple GPUs, reducing peak GPU memory requirements substantially.
+Advanced GPU Optimizations
+--------------------------
 
-For more information, refer to :ref:`sequential-parallelism`.
+When training on single or multiple GPU machines, Lightning offers a host of advanced optimizations to improve throughput, memory effeciency and model scaling.
```
Comment (carmocca) [inline docs/source/advanced/training_tricks.rst:155]:
```suggestion
When training on single or multiple GPU machines, Lightning offers a host of advanced optimizations to improve throughput, memory efficiency, and model scaling.
```

File: docs/source/advanced/advanced_gpu.rst [hunk 10]
```diff
@@ -0,0 +1,571 @@
+Advanced GPU Optimized Training
+===============================
+
+When training large models, fitting larger batch sizes or trying to increase throughput using multi-gpu compute, Lightning provides advanced optimized multi-gpu plugins to support these cases. For example if you'd like to train a large billion parameter transformer model, or to scale your batch size when training a semi-supervised learning model, using a Lightning optimized distributed training plugin will offer substantial improvements in GPU memory usage.
+
+Note that some of the extreme memory saving configurations will affect the speed of training. This Speed/Memory trade-off in most cases can be adjusted.
+
+Some of these memory efficient plugins rely on offloading onto other forms of memory, such as CPU RAM or NVMe. This means you can even see memory benefits on a **single GPU**, using a plugin such as :ref:`deepspeed-zero-stage-3-offload`.
```
Comment (carmocca) [inline docs/source/advanced/advanced_gpu.rst:8]: First paragraph seemed repeated, also typos

```suggestion
When training large models, fitting larger batch sizes, or trying to increase throughput using multi-GPU compute, Lightning provides advanced optimized distributed training plugins to support these cases and offer substantial improvements in memory usage.

Note that some of the extreme memory saving configurations will affect the speed of training. This Speed/Memory trade-off in most cases can be adjusted.

Some of these memory-efficient plugins rely on offloading onto other forms of memory, such as CPU RAM or NVMe. This means you can even see memory benefits on a **single GPU**, using a plugin such as :ref:`deepspeed-zero-stage-3-offload`.
```

File: docs/source/advanced/advanced_gpu.rst [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+
+Choosing an Advanced Distributed GPU Plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you would like to stick with PyTorch DDP, see :ref:`ddp-optimizations`.
+
+Unlike PyTorch Distributed Data Parallel (DDP) where the maximum trainable model size and batch size do not change with respect to the number of GPUs, memory optimized plugins can accommodate bigger model and larger batch as more GPUs are used. This means as you scale up the number of GPUs, you can reach the number of model parameters you'd like to train.
```
Comment (carmocca) [inline docs/source/advanced/advanced_gpu.rst:15]:
```suggestion
Unlike PyTorch's DistributedDataParallel (DDP) where the maximum trainable model size and batch size do not change with respect to the number of GPUs, memory-optimized plugins can accommodate bigger models and larger batches as more GPUs are used. This means as you scale up the number of GPUs, you can reach the number of model parameters you'd like to train.
```

File: docs/source/advanced/advanced_gpu.rst [hunk 12]
```diff
@@ -0,0 +1,571 @@
...
+
+Pre-training vs Fine-tuning
+"""""""""""""""""""""""""""
+
+When fine-tuning, we often use a magnitude less data compared to pre-training a model. This is important when choosing a distributed plugin as usually for pre-training, **where we are compute bound**.
```
Comment (carmocca) [inline docs/source/advanced/advanced_gpu.rst:20]:
```suggestion
When fine-tuning, we often use a magnitude less data compared to pre-training a model. This is important when choosing a distributed plugin as usually for pre-training, **where we are compute-bound**.
```

File: docs/source/advanced/advanced_gpu.rst [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+* For both fine-tuning and pre-training, use :ref:`deepspeed-activation-checkpointing` or :ref:`fairscale-activation-checkpointing` as the throughput degradation is not significant
+
+For example when using 128 GPUs, you can **pre-train** large 10 to 20 Billion parameter models using :ref:`deepspeed-zero-stage-2` without having to take a performance hit with more advanced optimized multi-gpu plugins.
+
+But for **fine-tuning** a model, you can reach 10 to 20 Billion parameter models using :ref:`deepspeed-zero-stage-3-offload` on a **single GPU**. This does come with a significant throughput hit, which needs to be weighed accordingly.
+
+When Shouldn't I use an Optimized Distributed Plugin?
+"""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Sharding techniques help when model sizes are fairly large; roughly 500M+ parameters is where we've seen benefits. However, in cases where your model is small (ResNet50 of around 80M Parameters) it may be best to stick to normal distributed training, unless you are using unusually large batch sizes.
```
Comment (carmocca) [inline docs/source/advanced/advanced_gpu.rst:36]:
```suggestion
Sharding techniques help when model sizes are fairly large; roughly 500M+ parameters is where we've seen benefits. However, in cases where your model is small (ResNet50 of around 80M Parameters) it may be best to stick to ordinary distributed training unless you are using unusually large batch sizes.
```

**Comment by SeanNaren** [comment]
Thanks so much guys, should've addressed all points (thanks @carmocca for cleaning up!)

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
